hedera-node/src/main/java/com/hedera/services/stream/RecordStreamFileWriter.java

```diff
@@ -0,0 +1,490 @@
+
+	/**
+<	 * the current record stream type;
+	 * used to obtain file extensions and versioning
+	 */
+	private final RecordStreamType streamType;
+
+	/**
+	 * a messageDigest object for digesting entire stream file and generating entire record stream file hash
+	 */
+	private final MessageDigest streamDigest;
+
+	/**
+	 * a messageDigest object for digesting metaData in the stream file and generating metaData hash.
+	 * Metadata contains:
+	 * record stream version || HAPI proto version || startRunningHash || endRunningHash || blockNumber, where
+	 * || denotes concatenation
+	 */
+	private final MessageDigest metadataStreamDigest;
+
+	/**
+	 * file stream and output stream for writing record stream data to file
+	 * */
+	private FileOutputStream stream = null;
+	private SerializableDataOutputStream dos = null;
+
+	/**
+	 * output stream for digesting metaData
+	 */
+	private SerializableDataOutputStream dosMeta = null;
+
+	/**
+	 * current runningHash before consuming the object added by calling {@link #addObject(T)} method
+	 */
+	private RunningHash runningHash;
+
+	/**
+	 * signer for generating signatures
+	 */
+	private final Signer signer;
+
+	/**
+	 * period of generating record stream files in ms
+	 */
+	private final long logPeriodMs;
+
+	/**
+	 * initially, set to be consensus timestamp of the first object;
+	 * start to write a new file when the next object's consensus timestamp is in different periods with
+	 * lastConsensusTimestamp;
+	 * update its value each time receives a new object
+	 */
+	private Instant lastConsensusTimestamp;
+
+	/**
+	 * if it is true, we don't write object stream file until the first complete window. This is suitable for streaming
+	 * after reconnect, so that reconnect node can generate the same stream files as other nodes after reconnect.
+	 *
+	 * if it is false, we start to write object stream file immediately. This is suitable for streaming after
+	 * restart, so that no object would be missing in the nodes' stream files
+	 */
+	private boolean startWriteAtCompleteWindow;
+
+	/**
+	 * The current record stream file that is being written.
+	 */
+	private File file;
+
+	/**
+	 * The file name of the current record stream file. Used for logs.
+	 */
+	private String fileNameShort;
+
+	/**
+	 * the path to which we write record stream files and signature files
+	 */
+	private final String dirPath;
+	private RecordStreamFile.Builder recordStreamFileBuilder;
+	private int recordFileVersion;
+
+	public RecordStreamFileWriter(
+			final String dirPath,
+			final long logPeriodMs,
+			final Signer signer,
+			final boolean startWriteAtCompleteWindow,
+			final RecordStreamType streamType
+	) throws NoSuchAlgorithmException {
+		this.dirPath = dirPath;
+		this.logPeriodMs = logPeriodMs;
+		this.signer = signer;
+		this.startWriteAtCompleteWindow = startWriteAtCompleteWindow;
+		this.streamType = streamType;
+		this.streamDigest = MessageDigest.getInstance(DigestType.SHA_384.algorithmName());
```

Comment:
This `DigestType.SHA_384` is repeated in a lot of places. We should move it to a static variable so it's defined once in case it needs to change in the future.

Reply:
[Fixed](https://github.com/hashgraph/hedera-services/pull/3461/commits/4cc3f212e7d35ef866331a8df180ba914b6d3ca7).